File: Day03/Part02/day03_02.java
import java.io.BufferedReader;
import java.io.FileReader;


public class day03_02 {

    public static void main (String[] args) {

        int target = -1;

        try {
            BufferedReader br = new BufferedReader(new FileReader("input.txt"));
            target = Integer.parseInt(br.readLine());
            br.close();
        } catch (Exception e) {
            e.printStackTrace();
        }

        // Find the biggest matrix we will need
        int maxWidth = findMaxWidth(target);
        System.out.println("Target: " + target);
        System.out.println("Max Width: " + maxWidth);

        // Build the starting matrix
        int[][] matrix = new int[maxWidth][maxWidth];

        // Keep track of current number, position, and current loop size
        int currentNumber = 1;
        int xPos = maxWidth / 2;
        int yPos = maxWidth / 2;
        int loopSize = 1;
        matrix[xPos][yPos] = currentNumber;

        // Keep adding numbers to matrix until we reach the target
        while (currentNumber < target) {
            loopSize += 2;
            // move cursor right to start
            yPos += 1;
            // right column
            for (int i = 0; i < loopSize - 1; i++) {
                matrix[xPos][yPos] = currentNumber = addNeighbors(xPos, yPos, matrix);
                if (currentNumber > target) { 
                    break; 
                }
                if (i < loopSize - 2) { 
                    xPos -= 1; 
                };
            };
            if (currentNumber > target) { 
                break; 
            };
            // top row
            yPos -= 1;
            for (int i = 0; i < loopSize - 1; i++) {
                matrix[xPos][yPos] = currentNumber = addNeighbors(xPos, yPos, matrix);
                if (currentNumber > target) { 
                    break; 
                }
                if (i < loopSize - 2) { 
                    yPos -= 1; 
                };
            };
            if (currentNumber > target) { 
                break; 
            };
            // left column
            xPos += 1;
            for (int i = 0; i < loopSize - 1; i++) {
                matrix[xPos][yPos] = currentNumber = addNeighbors(xPos, yPos, matrix);
                if (currentNumber > target) { 
                    break; 
                }
                if (i < loopSize - 2) { 
                    xPos += 1; 
                };
            };
            if (currentNumber > target) { 
                break; 
            };
            // bottom row
            yPos += 1; 
            for (int i = 0; i < loopSize - 1; i++) {
                matrix[xPos][yPos] = currentNumber = addNeighbors(xPos, yPos, matrix);
                if (currentNumber > target) { 
                    break; 
                }
                if (i < loopSize - 2) { 
                    yPos += 1; 
                };
            };
        };

        // Print the answer
        System.out.println("Answer: " + currentNumber);
            
    };

    private static int findMaxWidth(int t) {
        int size = 1;
        while (t > size*size) {
            size += 2;
        }
        return size;
    }

    private static int addNeighbors(int x, int y, int[][] matrix) {
        int runningTotal = matrix[x][y];
        runningTotal += matrix[x][y-1];
        runningTotal += matrix[x][y+1];
        runningTotal += matrix[x-1][y-1];
        runningTotal += matrix[x-1][y];
        runningTotal += matrix[x-1][y+1];
        runningTotal += matrix[x+1][y-1];
        runningTotal += matrix[x+1][y];
        runningTotal += matrix[x+1][y+1];
        return runningTotal;
    }

    private static void printMatrix(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            System.out.print("[ ");
            for (int k = 0; k < matrix[i].length; k++) {
                System.out.print(" " + matrix[i][k] + ", ");
            }
            System.out.print(" ]\n");
        }
    }

}

File: Day03/Part02/day03_02.py
def calculate_neighbors(x, y, working_grid):
    """ calculates and returns the total of neighboring squares """
    running_total = working_grid[x][y]
    running_total += working_grid[x][y+1]
    running_total += working_grid[x][y-1]
    running_total += working_grid[x-1][y+1]
    running_total += working_grid[x-1][y]
    running_total += working_grid[x-1][y-1]
    running_total += working_grid[x+1][y+1]
    running_total += working_grid[x+1][y]
    running_total += working_grid[x+1][y-1]
    return running_total


def print_grid(gridy):
    for g in gridy:
        print(g)


# get number from input file
target = int(open("input.txt", "r").read())
print("Target:", target)

# find max possible grid size
max_grid_size = 1
while target > max_grid_size*max_grid_size:
    max_grid_size += 2
    print(max_grid_size)

# create the grid
grid = [[0] * max_grid_size for i in range(max_grid_size)]

# keep track of the current number so we know when it exceeds the target
current_number = 1

# fill in first number
current_node = (max_grid_size//2, max_grid_size//2)
grid[current_node[0]][current_node[1]] = current_number

loop_size = 1

while current_number < target:

    loop_size += 2
    print("Loop Size:", loop_size)

    # start at node right of last updated
    current_node = (current_node[0], current_node[1] + 1)

    # right - move up right column
    for i in range(loop_size - 1):
        print("Current Node:", current_node)
        grid[current_node[0]][current_node[1]] = current_number = calculate_neighbors(current_node[0], current_node[1], grid)
        # break out early if ans found
        if current_number > target:
            break
        # dont move on last loop
        if i < loop_size - 2:
            current_node = (current_node[0] - 1, current_node[1])

    # break out early if ans found
    if current_number > target:
        break

    # top - move left on top row
    current_node = (current_node[0], current_node[1] - 1)
    for i in range(loop_size - 1):
        grid[current_node[0]][current_node[1]] = current_number = calculate_neighbors(current_node[0], current_node[1], grid)
        # break out early if ans found
        if current_number > target:
            break
        # dont move on last loop
        if i < loop_size - 2:
            current_node = (current_node[0], current_node[1] - 1)

    # break out early if ans found
    if current_number > target:
        break

    # left - move down left column
    current_node = (current_node[0] + 1, current_node[1])
    for i in range(loop_size - 1):
        grid[current_node[0]][current_node[1]] = current_number = calculate_neighbors(current_node[0], current_node[1], grid)
        # break out early if ans found
        if current_number > target:
            break
        # dont move on last loop
        if i < loop_size - 2:
            current_node = (current_node[0] + 1, current_node[1])

    # break out early if ans found
    if current_number > target:
        break

    # bottom - move right on bottom row
    current_node = (current_node[0], current_node[1] + 1)
    for i in range(loop_size - 1):
        grid[current_node[0]][current_node[1]] = current_number = calculate_neighbors(current_node[0], current_node[1], grid)
        # break out early if ans found
        if current_number > target:
            break
        # dont move on last loop
        if i < loop_size - 2:
            current_node = (current_node[0], current_node[1] + 1)


print_grid(grid)
print("Answer:", current_number)
